micropip/transaction.py

```diff
@@ -52,7 +52,17 @@ async def gather_requirements(
         for requirement in requirements:
             requirement_promises.append(self.add_requirement(requirement))
 
-        await asyncio.gather(*requirement_promises)
+        futures: list[asyncio.Future] = []
+        try:
+            for coro in requirement_promises:
+                futures.append(asyncio.ensure_future(coro))
+            await asyncio.gather(*futures)
+        except ValueError:
+            if not self.keep_going:
+                for future in futures:
+                    if not future.done():
+                        future.cancel()
```

Comment:
I haven't investigated how cancellation interacts with our event loop and I'm slightly worried that it could not interact well. But if this passes the tests then we should merge.

Reply:
Wait a second, I just find that this part seems to raises several `asyncio.exceptions.InvalidStateError: invalid state` exceptions

Reply:
@Zac-HD any tips on how to make my event loop support cancellation? Where to look who to ask etc?

Reply:
Anyways I'll open a separate issue about this.

Reply:
https://github.com/pyodide/pyodide/issues/4800

Reply:
Could you try using a TaskGroup? Maybe that will fix our problems.
https://docs.python.org/3/library/asyncio-task.html#asyncio.TaskGroup

Reply:
Done, I've refactored this section using `TaskGroup`, now all the futures should be gathered in one top-level `TaskGroup`.

Note that this makes the traceback looks a bit different because `TaskGroup` would group all the exceptions into a `ExceptionGroup`.

Reply:
The test fails because the raised error is not a `ValueError` but a `ExceptionGroup` containing a `ValueError`.

It's possible to recover the original ValueError by wrap the async with block with a try-except:

```py
    try:
        async with asyncio.TaskGroup() as tg:
            self.tg = tg  # only one task group from the top level
            for requirement in requirements:
                tg.create_task(self.add_requirement(requirement))
    except ExceptionGroup as e:
        raise e.exceptions[-1] from None
```